File: Invert_Binary_Tree.cpp
/*Invert a binary tree.

     4
   /   \
  2     7
 / \   / \
1   3 6   9
to
     4
   /   \
  7     2
 / \   / \
9   6 3   1 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        if(!root){ return NULL;}
        TreeNode* res = new TreeNode(root->val);
        if(root->left){
            res->right = invertTree(root->left);
        }
        else
        {
            res->right = NULL;
        }
        if(root->right){
            res->left = invertTree(root->right);
        }
        else
        {
            res->left = NULL;
        }
        return res;
    }
};

File: python/Invert_Binary_Tree.py
""" Invert a binary tree.

     4
   /   \
  2     7
 / \   / \
1   3 6   9
to
     4
   /   \
  7     2
 / \   / \
9   6 3   1 """
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def invertTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if root is None: return None
        res = TreeNode(root.val)
        if root.left != None:
            res.right = self.invertTree(root.left)
        if root.right != None:
            res.left = self.invertTree(root.right)
        return res    
